servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java

```diff
@@ -0,0 +1,694 @@
+                if (inetWatchers != null) {
+                    for (ARecordPublisher inetWatcher : inetWatchers) {
+                        inetWatcher.close0();
+                    }
+                }
+
+                // We will filter the null value out on the outer flatMap transformation.
+                return Single.succeeded(null);
+            }
+        })
+        // When the SRV record becomes unavailable we clean up state and return a Single.succeeded(null), but
+        // we don't want to propagate this downstream, so filter it out.
+        .filter(Objects::nonNull);
+    }
+
+    private void addSrvEntry0(final ARecordPublisher entry) {
+        assertInEventloop();
+
+        srvARecordMap.computeIfAbsent(entry.inetHost, k -> new ArrayList<>(2)).add(entry);
+    }
+
+    private void removeSrvEntry0(final ARecordPublisher entry) {
+        assertInEventloop();
+
+        final List<ARecordPublisher> inetWatchers = srvARecordMap.get(entry.inetHost);
+        if (inetWatchers == null) {
+            return;
+        }
+        inetWatchers.remove(entry);
+        if (inetWatchers.isEmpty()) {
+            srvARecordMap.remove(entry.inetHost);
+        }
+    }
+
+    @Override
+    public Completable onClose() {
+        return fromSource(closeCompletable);
+    }
+
+    @Override
+    public Completable closeAsync() {
+        return new SubscribableCompletable() {
+            @Override
+            protected void handleSubscribe(final CompletableSource.Subscriber subscriber) {
+                closeCompletable.subscribe(subscriber);
+                if (nettyIoExecutor.isCurrentThreadEventLoop()) {
+                    closeAsync0();
+                } else {
+                    nettyIoExecutor.asExecutor().execute(DefaultDnsClient.this::closeAsync0);
+                }
+            }
+        };
+    }
+
+    private void closeAsync0() {
+        assertInEventloop();
+
+        if (closed) {
+            return;
+        }
+        closed = true;
+        resolver.close();
+        ttlCache.clear();
+        closeCompletable.onComplete();
+    }
+
+    private void assertInEventloop() {
+        assert nettyIoExecutor.isCurrentThreadEventLoop() : "Must be called from the associated eventloop.";
+    }
+
+    private final class SrvRecordPublisher extends EntriesPublisher<HostAndPort> {
+        private final String serviceName;
+
+        private SrvRecordPublisher(String serviceName) {
+            this.serviceName = serviceName;
+        }
+
+        @Override
+        protected AbstractEntriesPublisherSubscription newSubscription(
+                final Subscriber<? super Iterable<ServiceDiscovererEvent<HostAndPort>>> subscriber) {
+            return new AbstractEntriesPublisherSubscription(subscriber) {
+                @Override
+                protected Future<DnsAnswer<HostAndPort>> doDnsQuery() {
+                    Promise<DnsAnswer<HostAndPort>> promise = ImmediateEventExecutor.INSTANCE.newPromise();
```

Comment:
This can just be a follow up but it looks like we have all the context here to actually do the A lookup when the SRV lookup finishes in netty itself. This will simplify the control flow by eliminating the `flatMap` on the SRV publisher and the whole mechanics of representing the DNS A record lookup as an infinite `Publisher`.

I am still not convinced that we need to keep listening to the A lookup on the SRV record result. It feels surprising to have both SRV record and a changing A record for each SRV record result.

Reply:
lets discuss as followup.

current goals are:
- re-use existing control flow/logic
- represent the DNS control flow as defined in the DNS SRV RFC [1]

[1] https://tools.ietf.org/html/rfc2782
> Target
        The domain name of the target host.  There MUST be one **or more**
        address records for this name,

Reply:
>    The domain name of the target host.  There MUST be one **or more**

To be clear; I am suggesting to do a one time lookup in netty for each SRV record entry which will give all A* records. 
The difference is that with the `ARecordPublisher` we are also listening to changes on that A record over time which seems unnecessary in this case.

Reply:
The only constraint from the SRV RFC is target can't be an alias, other than that it is a normal A* resolution (which has its own TTL and may change over time). If the DNS server provides the A* resolution in the `Additional` section of the SRV answer we may be able to avoid the secondary lookup, but depends upon the DNS server behavior (we can explore this optimization later).